gen/unmarshal.go

```diff
@@ -220,12 +220,18 @@ func (u *unmarshalGen) mapstruct(s *Struct) {
 		}
 		u.p.printf("\ncase \"%s\":", s.Fields[i].FieldTag)
 		u.p.printf("\nif validate && %s && \"%s\" < %s {", lastIsSet, s.Fields[i].FieldTag, last)
-		u.p.print("\nerr = &msgp.ErrNonCanonical{}")
+		u.p.print("\nerr = msgp.ErrNonCanonical(\"struct fields out of order\")")
 		u.p.printf("\nreturn")
 		u.p.print("\n}")
 		u.ctx.PushString(s.Fields[i].FieldName)
 		next(u, s.Fields[i].FieldElem)
 		u.ctx.Pop()
+		if ize := s.Fields[i].FieldElem.IfZeroExpr(); ize != "" && isFieldOmitEmpty(s.Fields[i], s) {
```

Comment:
This is not robust. If `IfZeroExpr()` returns empty string it will validate unconditionally